src/groupeddataframe/grouping.jl

```diff
@@ -366,62 +381,259 @@ _nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
 _ncol(df::AbstractDataFrame) = ncol(df)
 _ncol(x::Union{NamedTuple, DataFrameRow}) = length(x)
 
-function gen_fun(::Type{NT}, f2) where NT
-    get_val(::Val{T}) where {T} = T
+abstract type AbstractAggregate end
 
-    function(incols)
-        tup = map(f2) do p
-            nms = get_val(first(p))
-            if nms isa Tuple
-                res = last(p)(NamedTuple{nms}(map(c -> incols[c], nms)))
-            else
-                res = last(p)(incols[nms])
+struct Reduce{O, C, A} <: AbstractAggregate
+    op::O
+    condf::C
+    adjust::A
+end
+Reduce(f, condf=nothing) = Reduce(f, condf, nothing)
+
+check_aggregate(f::Any) = f
+check_aggregate(::typeof(sum)) = Reduce(Base.add_sum)
+check_aggregate(::typeof(prod)) = Reduce(Base.mul_prod)
+check_aggregate(::typeof(maximum)) = Reduce(max)
+check_aggregate(::typeof(minimum)) = Reduce(min)
+check_aggregate(::typeof(mean)) = Reduce(Base.add_sum, nothing, /)
+check_aggregate(::typeof(sum∘skipmissing)) = Reduce(Base.add_sum, !ismissing)
```

Comment:
Cool... I hadn't thought to introspect the type of the composition in this way - but this is a really clear way of expressing this!